packages/server/src/fhir/lookups/reference.ts

```diff
@@ -0,0 +1,118 @@
+import { Reference, Resource, ResourceType, SearchParameter } from '@medplum/fhirtypes';
+import { PoolClient } from 'pg';
+import { LookupTable } from './lookuptable';
+import { PropertyType, evalFhirPathTyped, getSearchParameters, isUUID, toTypedValue } from '@medplum/core';
+import { SelectQuery } from '../sql';
+import { compareArrays } from './util';
+
+/**
+ * The ReferenceTable class represents a set of lookup tables for references between resources.
+ * Each reference is represented as a separate row in the "<ResourceType>_References" table.
+ */
+export class ReferenceTable extends LookupTable<Reference> {
+  getTableName(resourceType: ResourceType): string {
+    return resourceType + '_References';
+  }
+
+  getColumnName(): string {
+    return '';
+  }
+
+  /**
+   * Returns false, because the table is never used for normal SearchParameter search.
+   * Instead, it is only used for JOINs between tables.
+   * @returns False
+   */
+  isIndexed(): boolean {
+    return false;
+  }
+
+  async indexResource(client: PoolClient, resource: Resource): Promise<void> {
+    const values = getSearchReferences(resource);
+    const existing = await getExistingValues(client, this.getTableName(resource.resourceType), resource.id as string);
+    if (compareArrays(values, existing)) {
+      // Nothing changed
+      return;
+    }
+    if (existing.length > 0) {
+      await this.deleteValuesForResource(client, resource);
+    }
+    await this.insertValuesForResource(client, resource.resourceType, values, true);
```

Comment:
If we call `deleteValuesForResource`, do we still need to pass `true` for `ignoreConflict`?  Did you find a case where the `ignoreConflict` option was necessary?

Reply:
I think the context got lost when I split this out from the original PR: https://github.com/medplum/medplum/pull/3251#discussion_r1387011984

Reply:
Hm - Ok, I think we should fix that either inside `insertValuesForResource` or elsewhere when building the list.  We should use a `Set<string>` to ensure there are no duplicate values.